File: runtime/schema/go/parse/parser.go
package parse

import (
	"encoding/json"
	"fmt"

	"github.com/parable-work/superschematic/ir"
	"gopkg.in/yaml.v3"
)

// Parser interprets an IR Schema to coerce, normalize, default, and
// scalar-parse runtime data maps. It is independent of validate so callers
// can choose to run parse, validate, or both. See the package doc for the
// per-field operation order.
type Parser struct {
	schema   *ir.Schema
	parseReg *ParseRegistry
	normReg  *NormalizeRegistry
	strict   bool
}

// Option configures a Parser during construction.
type Option func(*Parser)

// WithParseRegistry sets the registry of custom scalar parse functions.
// Defaults to an empty registry; HasCustomParse scalars without a matching
// entry simply skip the parse step.
func WithParseRegistry(r *ParseRegistry) Option {
	return func(p *Parser) {
		p.parseReg = r
	}
}

// WithNormalizeRegistry sets the registry of custom scalar normalize
// functions. Defaults to an empty registry; HasCustomNormalize scalars
// without a matching entry skip the normalize step.
func WithNormalizeRegistry(r *NormalizeRegistry) Option {
	return func(p *Parser) {
		p.normReg = r
	}
}

// WithStrict makes the default FromMap / ParseType / ParseInput entry points
// behave as their Strict variants (reject unknown fields, no coercion).
// Convenience for callers that pass a strict flag through configuration.
func WithStrict(strict bool) Option {
	return func(p *Parser) {
		p.strict = strict
	}
}

// New creates a Parser for the given schema. Both registries default to
// empty; pass [WithParseRegistry] / [WithNormalizeRegistry] with the
// DefaultParseRegistry / DefaultNormalizeRegistry constructors to wire the
// scalar-lib defaults.
func New(schema *ir.Schema, opts ...Option) *Parser {
	p := &Parser{
		schema:   schema,
		parseReg: NewParseRegistry(),
		normReg:  NewNormalizeRegistry(),
	}
	for _, opt := range opts {
		opt(p)
	}
	return p
}

// ParseType parses raw into a result map honouring typeName's TypeDef.
// Unknown fields are kept in lenient mode and rejected with
// {Validator: "unknown_field"} in strict mode (toggled via [WithStrict] or
// [Parser.ParseTypeStrict]).
func (p *Parser) ParseType(typeName string, raw map[string]any) (map[string]any, ValidationErrors) {
	return p.parseTypeWithStrict(typeName, raw, p.strict)
}

// ParseInput parses raw into a result map honouring inputName's TypeDef.
// Same lenient/strict semantics as ParseType.
func (p *Parser) ParseInput(inputName string, raw map[string]any) (map[string]any, ValidationErrors) {
	return p.parseInputWithStrict(inputName, raw, p.strict)
}

// ParseTypeStrict is ParseType with strict=true.
func (p *Parser) ParseTypeStrict(typeName string, raw map[string]any) (map[string]any, ValidationErrors) {
	return p.parseTypeWithStrict(typeName, raw, true)
}

// ParseInputStrict is ParseInput with strict=true.
func (p *Parser) ParseInputStrict(inputName string, raw map[string]any) (map[string]any, ValidationErrors) {
	return p.parseInputWithStrict(inputName, raw, true)
}

// FromMap is a name-agnostic helper that dispatches to ParseInput if a name
// matches an input definition, otherwise to ParseType.
func (p *Parser) FromMap(name string, raw map[string]any) (map[string]any, ValidationErrors) {
	if _, ok := p.schema.Inputs[name]; ok {
		return p.parseInputWithStrict(name, raw, p.strict)
	}
	return p.parseTypeWithStrict(name, raw, p.strict)
}

// FromMapStrict is FromMap with strict=true.
func (p *Parser) FromMapStrict(name string, raw map[string]any) (map[string]any, ValidationErrors) {
	if _, ok := p.schema.Inputs[name]; ok {
		return p.parseInputWithStrict(name, raw, true)
	}
	return p.parseTypeWithStrict(name, raw, true)
}

// ParseTypeJSON unmarshals data as JSON and parses it as typeName.
func (p *Parser) ParseTypeJSON(typeName string, data []byte) (map[string]any, ValidationErrors) {
	raw, errs := unmarshalJSON(data)
	if errs.HasErrors() {
		return nil, errs
	}
	return p.parseTypeWithStrict(typeName, raw, p.strict)
}

// ParseInputJSON unmarshals data as JSON and parses it as inputName.
func (p *Parser) ParseInputJSON(inputName string, data []byte) (map[string]any, ValidationErrors) {
	raw, errs := unmarshalJSON(data)
	if errs.HasErrors() {
		return nil, errs
	}
	return p.parseInputWithStrict(inputName, raw, p.strict)
}

// ParseTypeYAML unmarshals data as YAML and parses it as typeName.
func (p *Parser) ParseTypeYAML(typeName string, data []byte) (map[string]any, ValidationErrors) {
	raw, errs := unmarshalYAML(data)
	if errs.HasErrors() {
		return nil, errs
	}
	return p.parseTypeWithStrict(typeName, raw, p.strict)
}

// ParseInputYAML unmarshals data as YAML and parses it as inputName.
func (p *Parser) ParseInputYAML(inputName string, data []byte) (map[string]any, ValidationErrors) {
	raw, errs := unmarshalYAML(data)
	if errs.HasErrors() {
		return nil, errs
	}
	return p.parseInputWithStrict(inputName, raw, p.strict)
}

func (p *Parser) parseTypeWithStrict(typeName string, raw map[string]any, strict bool) (map[string]any, ValidationErrors) {
	td, ok := p.schema.Types[typeName]
	if !ok {
		errs := NewValidationErrors()
		errs.AddFieldError("", "type", fmt.Sprintf("unknown type %q", typeName))
		return nil, errs
	}
	return p.walk(td, raw, strict)
}

func (p *Parser) parseInputWithStrict(inputName string, raw map[string]any, strict bool) (map[string]any, ValidationErrors) {
	td, ok := p.schema.Inputs[inputName]
	if !ok {
		errs := NewValidationErrors()
		errs.AddFieldError("", "type", fmt.Sprintf("unknown input %q", inputName))
		return nil, errs
	}
	return p.walk(td, raw, strict)
}

func (p *Parser) walk(td *ir.TypeDef, raw map[string]any, strict bool) (map[string]any, ValidationErrors) {
	if raw == nil {
		raw = map[string]any{}
	}
	result := make(map[string]any, len(td.Fields))
	errs := NewValidationErrors()
	p.walkTypeDef(td, raw, result, errs, strict)
	return result, errs
}

func unmarshalJSON(data []byte) (map[string]any, ValidationErrors) {
	errs := NewValidationErrors()
	var raw map[string]any
	if err := json.Unmarshal(data, &raw); err != nil {
		errs.AddFieldError("", "json", err.Error())
		return nil, errs
	}
	if raw == nil {
		errs.AddFieldError("", "json", "expected JSON object payload")
		return nil, errs
	}
	return raw, errs
}

func unmarshalYAML(data []byte) (map[string]any, ValidationErrors) {
	errs := NewValidationErrors()
	var raw map[string]any
	if err := yaml.Unmarshal(data, &raw); err != nil {
		errs.AddFieldError("", "yaml", err.Error())
		return nil, errs
	}
	if raw == nil {
		errs.AddFieldError("", "yaml", "expected YAML mapping payload")
		return nil, errs
	}
	return raw, errs
}

File: runtime/schema/python/superschematic_schema_runtime/parse/parser.py
"""Runtime parse entry points.

Public surface: ``parse_type`` / ``parse_input`` (map -> coerced + normalized +
default-filled map), ``parse_type_json`` / ``parse_input_json`` (decode JSON
bytes/string then dispatch).
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any

from .. import ValidationError
from ..errors import ValidationErrors, append_field_error, new_validation_errors
from ..validation.resolver import resolve_schema
from ..validation.types import Schema, SchemaInput
from .registry import (
    ScalarNormalizeRegistry,
    ScalarParseRegistry,
    create_default_scalar_normalize_registry,
    create_default_scalar_parse_registry,
)
from .walk import WalkContext, walk_type_def


@dataclass
class ParseOptions:
    strict: bool = False
    parse_registry: ScalarParseRegistry | None = None
    normalize_registry: ScalarNormalizeRegistry | None = None


@dataclass
class ParseResult:
    data: dict[str, Any]
    errors: ValidationErrors


_default_parse_registry: ScalarParseRegistry | None = None
_default_normalize_registry: ScalarNormalizeRegistry | None = None


def _get_default_parse_registry() -> ScalarParseRegistry:
    global _default_parse_registry
    if _default_parse_registry is None:
        _default_parse_registry = create_default_scalar_parse_registry()
    return _default_parse_registry


def _get_default_normalize_registry() -> ScalarNormalizeRegistry:
    global _default_normalize_registry
    if _default_normalize_registry is None:
        _default_normalize_registry = create_default_scalar_normalize_registry()
    return _default_normalize_registry


def _ctx(schema: Schema, options: ParseOptions | None) -> WalkContext:
    opts = options if options is not None else ParseOptions()
    return WalkContext(
        schema=schema,
        parse_registry=opts.parse_registry if opts.parse_registry is not None else _get_default_parse_registry(),
        normalize_registry=opts.normalize_registry
        if opts.normalize_registry is not None
        else _get_default_normalize_registry(),
        strict=opts.strict,
    )


def _unknown_type_error(name: str, kind: str) -> ParseResult:
    errs = new_validation_errors()
    append_field_error(
        errs, "", ValidationError(validator="type", message=f'unknown {kind} "{name}"')
    )
    return ParseResult(data={}, errors=errs)


def _object_from_value(value: Any) -> dict[str, Any]:
    return value if isinstance(value, dict) else {}


def parse_type(
    schema: SchemaInput,
    type_name: str,
    data: Any,
    options: ParseOptions | None = None,
) -> ParseResult:
    """Parse ``data`` against ``schema.types[type_name]``. Map-in/map-out;
    applies defaults, coerces primitives, runs normalize/parse hooks.
    """
    resolved = resolve_schema(schema)
    td = resolved.types.get(type_name)
    if td is None:
        return _unknown_type_error(type_name, "type")
    out, errs = walk_type_def(_ctx(resolved, options), td, _object_from_value(data))
    return ParseResult(data=out, errors=errs)


def parse_input(
    schema: SchemaInput,
    input_name: str,
    data: Any,
    options: ParseOptions | None = None,
) -> ParseResult:
    """Parse ``data`` against ``schema.inputs[input_name]``."""
    resolved = resolve_schema(schema)
    td = resolved.inputs.get(input_name)
    if td is None:
        return _unknown_type_error(input_name, "input")
    out, errs = walk_type_def(_ctx(resolved, options), td, _object_from_value(data))
    return ParseResult(data=out, errors=errs)


def parse_type_map(
    schema: SchemaInput,
    type_name: str,
    data: dict[str, Any],
    options: ParseOptions | None = None,
) -> ParseResult:
    return parse_type(schema, type_name, data, options)


def parse_input_map(
    schema: SchemaInput,
    input_name: str,
    data: dict[str, Any],
    options: ParseOptions | None = None,
) -> ParseResult:
    return parse_input(schema, input_name, data, options)


def _decode_json(payload: str | bytes) -> tuple[Any, ValidationErrors]:
    errors = new_validation_errors()
    if isinstance(payload, bytes):
        try:
            text = payload.decode("utf-8")
        except UnicodeDecodeError as exc:
            append_field_error(errors, "", ValidationError(validator="json", message=str(exc)))
            return None, errors
    else:
        text = payload
    try:
        return json.loads(text), errors
    except json.JSONDecodeError as exc:
        append_field_error(errors, "", ValidationError(validator="json", message=str(exc)))
        return None, errors


def parse_type_json(
    schema: SchemaInput,
    type_name: str,
    payload: str | bytes,
    options: ParseOptions | None = None,
) -> ParseResult:
    value, errs = _decode_json(payload)
    if errs:
        return ParseResult(data={}, errors=errs)
    if not isinstance(value, dict):
        e = new_validation_errors()
        append_field_error(
            e, "", ValidationError(validator="json", message="expected JSON object payload")
        )
        return ParseResult(data={}, errors=e)
    return parse_type(schema, type_name, value, options)


def parse_input_json(
    schema: SchemaInput,
    input_name: str,
    payload: str | bytes,
    options: ParseOptions | None = None,
) -> ParseResult:
    value, errs = _decode_json(payload)
    if errs:
        return ParseResult(data={}, errors=errs)
    if not isinstance(value, dict):
        e = new_validation_errors()
        append_field_error(
            e, "", ValidationError(validator="json", message="expected JSON object payload")
        )
        return ParseResult(data={}, errors=e)
    return parse_input(schema, input_name, value, options)


__all__ = [
    "ParseOptions",
    "ParseResult",
    "parse_input",
    "parse_input_json",
    "parse_input_map",
    "parse_type",
    "parse_type_json",
    "parse_type_map",
]
